cpu/efm32/periph/timer_series01.c

```diff
@@ -47,6 +50,73 @@
  */
 static timer_isr_ctx_t isr_ctx[TIMER_NUMOF];
 
+/* Bitmask storing which channel are periodic and which not. Only use
+ * getter/setter functions to access */
+static uint8_t is_periodic_mask[TIMER_NUMOF];
+/* Variable holding if the top value of the timer has been modified and by
+ * which. Only use getter/setter functions to access */
+static uint8_t modified_top_mask[TIMER_NUMOF];
+
+/* Check if the given channel on the given timer is periodic */
+static bool is_periodic(tim_t dev, int chan)
+{
+    assume((unsigned)chan < 8);
+    assume((unsigned)dev < TIMER_NUMOF);
+
+    return is_periodic_mask[dev] & (1U << chan);
+}
+
+/* Mark the given channel on the given timer as periodic */
+static void mark_periodic(tim_t dev, int chan)
+{
+    assume((unsigned)chan < 8);
+    assume((unsigned)dev < TIMER_NUMOF);
+
+    atomic_fetch_or_u8(&is_periodic_mask[dev], 1U << chan);
+}
+
+/* Mark the given channel on the given timer as on-shot timer */
+static void mark_one_shot(tim_t dev, int chan)
+{
+    assume((unsigned)chan < 8);
+    assume((unsigned)dev < TIMER_NUMOF);
+
+    atomic_fetch_and_u8(&is_periodic_mask[dev], ~(1U << chan));
+}
+
+/* Return the number of the channel that modified the top value of the given
+ * timer, or -1 if the top value is MAX */
+static int get_chan_which_modified_top_value(tim_t dev)
+{
+    assume((unsigned)dev < TIMER_NUMOF);
+    if (modified_top_mask[dev] == 0) {
+        return -1;
+    }
+
+    return modified_top_mask[dev] & 0x7f;
+}
+
+/* Mark the top value of the given timer as modified by the given channel */
+static void set_chan_which_modified_top_value(tim_t dev, int channel)
+{
+    assume((unsigned)dev < TIMER_NUMOF);
+    modified_top_mask[dev] = 0x80 | channel;
```

Comment:
This code suggests that `channel` is a power of two? While the code further above e.g. `atomic_fetch_or_u8(&is_periodic_mask[dev], 1U << chan);` suggests it is "the power itself".
2^3 <-> 3.
Please un-confuse me :face_with_spiral_eyes:

Reply:
I'll add a comment.

The implementation makes sure at most one periodic channel has `RESET_ON_MATCH` set and emulate that behavior by modifying the timers top value.

Since there is only ever one channel anyway, this is not a bitmask but just holds the numeric value of the channel. The most significant bit is set if this holds a value (to distinguish "channel 0 is the current reset-on-match" channel from "no reset on match"). But that was premature optimization, initially assumed that it would be smart to have `0` as the inital value so that I can rely on it to be initialized when `.bss` is cleared. But since the timer can be reinitialized anyway (and is currently explicitly writing a zero to it just in case it is a reinitialization), I could also just use `0xFF` as guard value and get easier to read C code and smaller machine code.

I'll adapt and document.